### ISSUE TITLE
Requesting definition of versioning scheme

### ISSUE THREAD

**Issue by cxw42** [editorconfig/editorconfig#395]
This is in reference to https://github.com/editorconfig/editorconfig-vim/pull/119#discussion_r280008712 and is related to #356 and https://github.com/editorconfig/editorconfig-core-test/issues/14 .

Would one of the team members please create a wiki page or other document defining the project's versioning conventions?  (Or point me to it if it already exists and I am missing something!)  I can't find it on docs.editorconfig.org or in the Wiki in this repo.  

What I think I know so far:

- The core tests carry a version number, which is the version number of the specification (https://github.com/editorconfig/editorconfig-core-test/issues/14).
- The C core uses a similar version number ([here](https://github.com/editorconfig/editorconfig-core-c/blob/master/CMakeLists.txt#L35)).  `PROJECT_VERSION*` from CMakeLists.txt  are copied to `EC_VERSION*` in src/config.h.in.
- The [core tests](https://github.com/editorconfig/editorconfig-core-test/blob/master/cli/CMakeLists.txt#L31) require the core to return a three-component version number.  
- v0.12.2 is the PyPI version of editorconfig-core-py.  
- v0.12 is the latest deployed version of editorconfig-core-test.  

Does this mean that plugins should use the core-test version as their MAJOR.MINOR, but have their own PATCH?  What is the plan for the version numbering?  Thanks for helping me understand!

If there is no specific plan, would someone please write one?  I would personally suggest:

- Using semver for all repos
- Defining the version of the specification as the version of editorconfig-core-test
- Bumping editorconfig-core-test to v1.0.0, since the project is increasingly widely used
- Letting each core or plugin use its own versioning scheme for releases, independent of the specification/core-test version it supports.

**Comment by ppalaga** [comment]
> * Using semver for all repos

Please define "all repos".

> * Defining the version of the specification as the version of editorconfig-core-test

+1 as long as we do not have any other formal spec, such as grammar.

> * Bumping editorconfig-core-test to v1.0.0, since the project is increasingly widely used

-1 unless there is a breaking change

> Letting each core or plugin use its own versioning scheme for releases, independent of the specification/core-test version it supports.

+1

**Comment by cxw42** [comment]
@ppalaga Edited to clarify.

Re. the spec version bump to 1.0.0, semver [permits](https://semver.org/#spec-item-4) and [encourages](https://semver.org/#how-do-i-know-when-to-release-100) bumping to 1.0.0 when software is being used in production, so that downstream can begin relying on semver guarantees.  EditorConfig is definitely being used in production :D .  I do know that such a bump, and a commitment to semver, would require a Board vote.

**Comment by xuhdev** [comment]
Currently our scheme is already semver: The minor indicates a change in format support (from a core library perspective), the patch indicates software bug fixes. We have never invoked the major yet, so how it will be used is still unclear. I personally would leave it for incompatible changes, but IMO we can start a voting when it becomes relevant. For now, I do not see a compelling reason to upgrade to `v1.x.x`.

**Comment by xuhdev** [comment]
In fact, I think this is due to #360 . Due to an old voting thread #361 , I'll soon open a new repo with a basic document template.

**Comment by cxw42** [comment]
@xuhdev Thanks for the additional info!  What I think I am hearing is the following; please let me know if I am understanding this correctly.

- When the project started, there was only one core: the C core ([initial commit](https://github.com/editorconfig/editorconfig-core-c/commit/45142061d9afb09a66c300f72c6daee7a22f3a27) 2011-10-22).  The specification was whatever the C core accepted.
- The version number accepted by `editorconfig -b <ver>` is the specification version.
- At present, the specification version still matches the C core version.
- The `editorconfig-core-test` version lags behind the C core version because currently only the semver minor is used to indicate format changes.  The C core uses the semver patch for bugfixes.

If the above is correct, my request would be that the Board:

- officially define the version number for `editorconfig -b` / `-v` / `--version` purposes as the `editorconfig-core-test` version, so that the cores and plugins are decoupled from the specification
- bump `editorconfig-core-test` to `v0.13` to mark this change
- update the core tests (or accept a PR) [here](https://github.com/editorconfig/editorconfig-core-test/blob/b8f30cfea1d1c1cd46219280ad4eb27780d21f7b/cli/CMakeLists.txt#L29) to permit major.minor version numbers (no patch)
- officially permit plugins and cores to use version numbers that are different from `core-test`.
 - For example, the Vimscript core might respond to `-v` with

        EditorConfig Vimscript core v1.0.0 - Specification Version 0.13

What say you?

**Comment by xuhdev** [comment]
I think your understanding is correct. I personally agree with you that there should be an additional version number solely marking software changes. I don't think we have been really using the tags in `editorconfig-core-test`, but nonetheless we should have updated it (and I have just done so).

Let me start the voting after creating the basic spec template, so what is at stake would be much less confusing.

**Comment by cxw42** [comment]
@xuhdev Sounds good - thanks!

**Comment by cxw42** [comment]
Needed for editorconfig/editorconfig-vim#143 - what news?  Thanks!

**Comment by florianb** [comment]
I vote for yes in https://github.com/editorconfig/editorconfig-vote/issues/11,

but i also urge to aim for implementation of

- SemVer: in the many levels of moving parts in the EditorConfig universe it may be useful if not critical to be able to duistinguish between feature-implementations and patches.
- An optional global version property in the EditorConfig: we must provide assurances that EditorConfig behaves as intended by the user.

**Comment by cxw42** [comment]
@xuhdev Would you please open a vote on this question per @florianb ? 

> Should the version number of the specification (== the version number of editorconfig-core-test) follow the Semantic Versioning 2.0.0 standard (https://semver.org)?

https://github.com/editorconfig/editorconfig-vote/issues/11 approved changing the way we version the specification.  However, that vote didn't require any specific format of version.

Thanks!

**Comment by xuhdev** [comment]
Yes. Could you open a new vote? Just click on "New Issue" at editorconfig/editorconfig-vote and there is a template guiding you what to put in.

**Comment by xuhdev** [comment]
OK, let me do it.

**Comment by xuhdev** [comment]
@cxw42 Could you give a more detailed description of the question that you would like to have a vote on? semver.org is fairly long and I'm not sure whether you meant adopting it as a whole.

**Comment by florianb** [comment]
@xuhdev: I would propose to "adopt" it as whole. So that:

1. The Spec allows the differentiation between breaking changes (major version), non-breaking feature additions (like a new rule as minor-version) and patches/corrections (without any expected changes to functionality) to the spec itself (patch version).
2. An implementing EditorConfig-tool can display its current supported EC spec version, so that users are able recognize which EC-version is actually supported.

So f.e. the spec could be `1.4.2` (where in four iterations new non-breaking rule functionality was added and two subsequent mistakes were ironed out from the spec-documentation).

A implementing tool could then align f.e. to the version `1.4.1`, where the developer aligned with the spec version `1.4` and already had one patch released (in the tools scope).

---

Since we also have some "interpreters" which can't deal with some specific rules i could also think of having an indicator for "intentional missing rules", f.e. (i'm using the build-tag for that) like this:

`1.4.1+miss.2.6`

Which could indicate that this interpreter aligns with spec number `1.4` but intentionally misses the rules 2 and 4 (yeah we would also need to number our rules).

Adding support for a rule later on would be treated as a patch, f.e. the dev adds support for rule number 2, the new version number would become:

`1.4.2+miss.6`

This whole thing would also add the possibility to tailor the core-tests run against the implementation by the build-tag.

**Comment by cxw42** [comment]
@xuhdev I also understood this as adopting semver as a whole for the spec==core-test version numbers.

@florianb I agree that a _plugin_ that cannot implement certain rules should document that fact.  However, I do not believe semver build metadata should be used for that purpose:

- A _core_ should pass all the tests for the spec version it conforms to, even if not all plugins can make use of all the information the core provides.
- The very name "build metadata" creates expectations about what that field of the number will contain.  Lists of unsupported rules seem to me unrelated to the SHA, date, or other attributes of the build.
- Build metadata is ignored for version comparison.  Therefore, any change in functionality must change another component of the number.  What if a developer changed the set of supported rules but forgot to change one of the x.y.z?  We'd have two versions that compared identically but had different functionality.

I agree with recording the information, but I think it should not be shoe-horned into the version number.  What say you?

**Comment by florianb** [comment]
@cxw42 - thanks for the feedback.

- `core-test` might in fact be the wrong reference since i thought more of some kind of compliance test for tools around EC.
- I think of "build metadata" in a less dogmatic way and would also allow to interpret it as some kind of user defined metadata, which this could be. But anyway - i really don't insist on "having" this.

Having build-metadata being ignored for version comparison is no problem in my opinion. Missing the update of the version-number is a faulty situation and must be fixed anyways. Changes in the metadata should always being reflected in the patch level.

Let's say a developer finds out the implementation of rule *n* is faulty and adds it to the list of intentional left out rules by adding `+miss.n`. The new version of the plugin, tool, whatever now must be propagated to have people rely on a stable version again. This should be done by raising the patch level by one.

My whole argumentation about Semver is in fact driven by some kind of wish for more explicity. A user facing any unexpected behavior in her EC-toolchain should be able to  match her expectations at first glance.

If she goes for a specific rule number 25 on the EC-website which is explicitly available since 1.12 and she takes a look at her toolchain which shows compatibility to 1.12+miss.25 she may be able to quickly identify the root source of her unmet expectations.

**Comment by cxw42** [comment]
@florianb Managing user expectations is very important :) .  Could we have the vote on semver for core-test==specification first, then follow up in a separate issue with the wider compliance-test and labeling questions?

**Comment by florianb** [comment]
@cxw42 sure - you're right - it is a subsequent enhancement once we have SemVer... :smiley:

**Comment by cxw42** [comment]
@florianb thanks!

@xuhdev Do you have the information you need for the vote, or do you need more?  Please let me know either way.

**Comment by xuhdev** [comment]
@cxw42 Done. Turns out https://github.com/editorconfig/editorconfig/issues/395#issuecomment-628557400  is the most precise and concise description. I ducked myself into all the conversations :smile:

**Comment by florianb** [comment]
Thank you! :smiley:

**Comment by ffes** [comment]
I will vote "yes" but do think only MAJOR.MINOR should match. There are too many reasons the patch level can change.

**Comment by florianb** [comment]
good point, i completely agree. I understood it like we're going to define how to use Semver in a separate vote. Hopefully.

**Comment by cxw42** [comment]
The vote was affirmative!  https://github.com/editorconfig/editorconfig-vote/issues/12#issuecomment-633712170

@florianb I agree we should at least add some guidance about major/minor/patch bumps.  I have some initial thoughts I will add later --- time for me to log off for the day :) .

(BTW I sent you a LinkedIn request, but it wouldn't let me add a note for some reason.  If you have an invitation with someone of the same name, it's me :D .)

**Comment by florianb** [comment]
Thank you very much @cxw42 - i will be happy to help working on those!

(Thanks a lot for connecting - LinkedIn's UI is sometimes wonderful... :smile: )

**Comment by cxw42** [comment]
Initial thoughts:

- semver exists to permit downstream packages to choose "flexible (but not too flexible) dependency specifications" (semver v2).
- The downstream flow here is spec -> core-test -> core -> plugin -> editor.
- In software packages, non-breaking changes often add flexibility.  The editorconfig specification is different in that additions to the spec often remove flexibility by more clearly stating what is required.

Easy cases (in my opinion):
- Any change in the spec (and corresponding change in the tests) which would cause a core to change from passing a test to failing the same test is semver-major.
- Any change in the spec that permits a core to accept a wider range of input is semver-minor.

Medium difficulty:
- If the core tests fall out of sync with the spec, a fix to the core-tests requires a semver-patch bump.  Since the spec and core-tests are versioned the same, that would mean the spec would also get a semver-patch bump.

Hard choice:
- Suppose that:
  - A new test is added; and
  - A core that previously passed all the tests might or might not pass the new test.

  Is that change semver-minor or semver-major?  I think it might be semver-major, since it has the potential to break downstream.  I admit this might eventually lead to spec version 42.0.0 :D .  However, [the semver spec points out](https://semver.org/#if-even-the-tiniest-backwards-incompatible-changes-to-the-public-api-require-a-major-version-bump-wont-i-end-up-at-version-4200-very-rapidly) that this is not inherently bad.

What say you all?

**Comment by florianb** [comment]
Thank you @cxw42 - i try to recap that a bit and add my thoughts on it:

### MAJOR

- Additions, updates or removals to or from the SPEC that break any CORE test

### MINOR

- Additions, updates or removals to or from the SPEC that conclude an addition to the CORE as well.

### PATCH

- Any change to the given implementation or test-suite to align better to the intention of the specifying LAYER.

---

1. I would use MAJOR.MINOR to express the alignment of a LAYER to a specific combination of abilities.
2. I agree - given we would have `*` changes, we would have a MAJOR version number of 42. But i think this is unlikely. Once we redefine the SPEC it should be in a matter, allowing us extending it without breaking any rules.

#### The "LAYER"

I'd also like to introduce the label "Layer", because i think we should open the nomenclature to a more versatile and workflow dependent use of EditorConfig.

The core will remain the versatile EditorConfig-parser but i guess there is a strong demand to provide other toolings explicit access to the specification as well:

spec > core_test > core > editor > fixer² > linter

²) fixer might also reference to other source-changing technologies like git's autocrlf.

And to be honest i'd also like to have a small set of vision-statements allowing us to align our work on the specification to a higher goal, without the necessity of reoccurring discussions if any demand may be part of the EditorConfig or not:

vision > spec > core_test > core > editor > fixer > linter

I'm happy to hear your thoughts!

*) = 42

**Comment by ppalaga** [comment]
> ### MINOR
> * Additions, updates or removals to or from the SPEC that conclude an addition to the CORE as well.

Could you please rephrase "conclude an addition to the CORE"? I have troubles to figure out what it is supposed to mean.

> spec > core_test > core > editor > fixer² > linter

What is the difference between fixer and linter?

> And to be honest i'd also like to have a small set of vision-statements allowing us to align our work on the specification to a higher goal, without the necessity of reoccurring discussions if any demand may be part of the EditorConfig

+1!

**Comment by florianb** [comment]
Thanks @ppalaga  - i rephrased it - i'm sorry for the bad phrasing. :smile: 

Fixer = actually modifying sources
Linter = measuring the compliance of sources

I once had the idea it would be useful to distinguish between the two to allow us specifying explicit expected behavior for one or another category of tools.

**Comment by ppalaga** [comment]
Thanks, for the clarifications. 

> Fixer = actually modifying sources
> Linter = measuring the compliance of sources

I tend to understand checking and fixing as two operation modes of a linter. Fixing code usually depends on the checking code. But, yes, the distinction is useful, however we call it. BTW, "Formatter" perhaps sounds nicer than "Fixer".

**Comment by florianb** [comment]
Good point - thanks! I replaced "fixer" by "formatter". Fixer came initially from the `--fix` flag of eslint.

**Comment by cxw42** [comment]
@mipo256 your https://github.com/editorconfig/editorconfig/issues/280#issuecomment-2743747540 relates to this --- MUST vs. SHOULD conformance can be tied into version scheme if desired.
